File: 贪心/763划分字母区间/763.划分字母区间.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=763 lang=java
 *
 * [763] 划分字母区间
 */

// @lc code=start
class Solution {
    public List<Integer> partitionLabels(String s) {
        int[] codeArr = new int[26];
        int strLen = s.length();

        for (int i = 0; i < strLen; i++) {
            codeArr[s.codePointAt(i) - 'a'] = i;
        }

        List<Integer> list = new ArrayList<Integer>();
        int end =0, start = 0;
        for(int i = 0; i < strLen; i++) {
            end = Math.max(end, codeArr[s.codePointAt(i) - 'a']);
            if (i == end) {
                list.add(end - start +1);
                start = end +1;
            }
        }
        
        return list;
    }
}
// @lc code=end

File: 贪心/763划分字母区间/763.划分字母区间.js
/*
 * @lc app=leetcode.cn id=763 lang=javascript
 *
 * [763] 划分字母区间
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number[]}
 */
var partitionLabels = function(s) {
  const len = s.length;
  const codeArr = new Array(26);
  const codePointA = "a".codePointAt(0);

  for(let i = 0; i < len; i++) {
      codeArr[s.codePointAt(i)-codePointA] = i;
  }
  let end = start = 0;
  const partition = [];
  for(let i = 0; i < len; i++) {
      end = Math.max(end, codeArr[s.codePointAt(i)-codePointA]);
      if(i == end) {
          partition.push(end-start+1);
          start = end +1;
      }
  }
  return partition;
};
// @lc code=end

